### DISCUSSION TITLE
Look into customizing Formtastic's HTML or replacing it

### DISCUSSION THREAD

**Discussion by javierjulio** [activeadmin/activeadmin#7875] [General]
Its been quite painful to style Formtastic's HTML in #3862 that I'd like to update it to use a simpler component based approach like you see in Bootstrap or Foundation. I very much like Bootstrap's approach of using a `form-control` class instead of targeting HTML elements and applying defaults. Further styling fieldsets is incredibly painful across browsers. We have a panel component in #3862 which we should be reusing there. Or just using a different presentation for form headers.

We can customize the HTML output but it would require that we subclass [each input class](https://github.com/justinfrench/formtastic/tree/master/lib/formtastic/inputs) and override the `input_wrapping` method to use a new approach. Although the problem is that the [fieldset_wrapper.rb helper](https://github.com/justinfrench/formtastic/blob/master/lib/formtastic/helpers/fieldset_wrapper.rb) generates not just the `<fieldset>` element but the `<ol>` as well which isn't good.

Further digging shows that `form_builder.rb` actually opens up the `Formtastic::Inputs::Base` module to override the `input_wrapping` method which for me was unexpected.

A good idea which could be quite challenging is switching to simple_form which is maintained and better documented. It has easier way of namespacing and customizing without having to subclass everything. It was built with customization in mind.

**Comment by scarroll32** [comment]
+1 for moving to simple_form

**Comment by xn** [comment]
+1 for moving to simple_form

**Comment by scarroll32** [comment]
Any thoughts on the possibility of moving to simple_form?

**Comment by javierjulio** [comment]
@seanfcarroll yes I think it would be great. Certainly better than formtastic as far as customization goes which would be beneficial. I haven't tried integrating it though so I don't know what to expect but I checked out what it offered and to see if it was being maintained. It would be valuable to experiment integrating simple_form but I figure it would be quite challenging.

**Comment by julioag-rmd** [comment]
It's 2025 and given that formtastic seems to be (semi?)abandoned by the amount of work needed to make it work with Rails 8.1 + Devise5, maybe this is worth revisiting?

The migration work surely is daunting, maybe LLM tools can make it easier to get around?
